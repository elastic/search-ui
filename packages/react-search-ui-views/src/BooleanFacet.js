import PropTypes from "prop-types";
import React from "react";

import { FacetValue, FilterValue } from "./types";
import { appendClassName } from "./view-helpers";

function BooleanFacet({
  className,
  label,
  options,
  onChange,
  onRemove,
  values
}) {
  const trueCount = options.find(option => option.value === "true").count;
  const isSelected = values.includes("true");

  const apply = () => onChange("true");
  const remove = () => onRemove("true");
  const toggle = () => {
    isSelected ? remove() : apply();
  };

  return (
    <fieldset className={appendClassName("sui-facet", className)}>
      <legend className="sui-facet__title">{label}</legend>
      <div className="sui-boolean-facet">
        <div className={"sui-boolean-facet__option-input-wrapper"}>
          <label className="sui-boolean-facet__option-label">
            <div className="sui-boolean-facet__option-input-wrapper">
              <input
                className="sui-boolean-facet__checkbox"
                type="checkbox"
                onChange={toggle}
              />
              <span className="sui-boolean-facet__input-text">{label}</span>
            </div>
            <span className="sui-boolean-facet__option-count">{trueCount}</span>
          </label>
        </div>
      </div>
    </fieldset>
  );
}

BooleanFacet.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string.isRequired,
  onRemove: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(FacetValue).isRequired,
  values: PropTypes.arrayOf(FilterValue).isRequired,
  onChange: PropTypes.func.isRequired
};

export default BooleanFacet;
