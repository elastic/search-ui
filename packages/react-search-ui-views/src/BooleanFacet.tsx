import React from "react";
import { FacetViewProps } from "./types";

import { appendClassName } from "./view-helpers";

function BooleanFacet({
  className,
  label,
  options,
  onChange,
  onRemove,
  values
}: FacetViewProps) {
  const trueOptions = options.find((option) => option.value === "true");
  if (!trueOptions) return null;
  const isSelected = values.includes("true");

  const apply = () => onChange("true");
  const remove = () => onRemove("true");
  const toggle = () => {
    isSelected ? remove() : apply();
  };

  return (
    <fieldset className={appendClassName("sui-facet", className)}>
      <legend className="sui-facet__title">{label}</legend>
      <div className="sui-boolean-facet">
        <div className={"sui-boolean-facet__option-input-wrapper"}>
          <label className="sui-boolean-facet__option-label">
            <div className="sui-boolean-facet__option-input-wrapper">
              <input
                data-transaction-name={`facet - ${label}`}
                className="sui-boolean-facet__checkbox"
                type="checkbox"
                onChange={toggle}
                checked={isSelected}
              />
              <span className="sui-boolean-facet__input-text">{label}</span>
            </div>
            <span className="sui-boolean-facet__option-count">
              {trueOptions.count}
            </span>
          </label>
        </div>
      </div>
    </fieldset>
  );
}

export default BooleanFacet;
