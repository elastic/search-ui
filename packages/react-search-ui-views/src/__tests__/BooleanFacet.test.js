import React from "react";
import BooleanFacet from "../BooleanFacet";
import { shallow, mount } from "enzyme";

const params = {
  label: "A Facet",
  onRemove: jest.fn(),
  onChange: jest.fn(),
  options: [
    {
      value: "true",
      count: 10
    }
  ],
  values: []
};

it("renders", () => {
  const wrapper = shallow(<BooleanFacet {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const wrapper = shallow(
    <BooleanFacet className={customClassName} {...params} />
  );
  const { className } = wrapper.props();
  expect(className.includes(customClassName)).toBe(true);
});

it("onChange is called on click", () => {
  const wrapper = mount(<BooleanFacet {...params} />);

  wrapper.find("input").simulate("change");
  expect(params.onChange).toHaveBeenCalledTimes(1);
});

it("onRemove is called on click", () => {
  const wrapper = mount(
    <BooleanFacet
      {...{
        ...params,
        values: ["true"]
      }}
    />
  );

  wrapper.find("input").simulate("change");
  expect(params.onRemove).toHaveBeenCalledTimes(1);
});

it("will not render when there are no true options", () => {
  params.options = [];
  const wrapper = shallow(<BooleanFacet {...params} />);
  expect(wrapper).toMatchSnapshot();
});
