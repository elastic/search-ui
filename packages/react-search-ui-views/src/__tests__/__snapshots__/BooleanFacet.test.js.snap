// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<fieldset
  className="sui-facet"
>
  <legend
    className="sui-facet__title"
  >
    A Facet
  </legend>
  <div
    className="sui-boolean-facet"
  >
    <div
      className="sui-boolean-facet__option-input-wrapper"
    >
      <label
        className="sui-boolean-facet__option-label"
      >
        <div
          className="sui-boolean-facet__option-input-wrapper"
        >
          <input
            className="sui-boolean-facet__checkbox"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="sui-boolean-facet__input-text"
          >
            A Facet
          </span>
        </div>
        <span
          className="sui-boolean-facet__option-count"
        >
          10
        </span>
      </label>
    </div>
  </div>
</fieldset>
`;
